pkl-core/src/main/java/org/pkl/core/stdlib/test/report/SimpleReport.java

```diff
@@ -17,107 +17,137 @@
-    var builder = new StringBuilder();
+    var builder = new TextFormattingStringBuilder(useColor);
 
     builder.append("module ").append(results.moduleName()).append("\n");
 
     if (results.error() != null) {
       var rendered = results.error().exception().getMessage();
       appendPadded(builder, rendered, "  ");
-      builder.append('\n');
+      builder.appendLine();
     } else {
       reportResults(results.facts(), builder);
       reportResults(results.examples(), builder);
     }
 
-    if (results.isExampleWrittenFailure()) {
-      builder.append(results.examples().totalFailures()).append(" examples written\n");
-      writer.append(builder);
-      return;
-    }
-
-    builder.append(results.failed() ? "❌ " : "✅ ");
-
-    var totalStatsLine =
-        makeStatsLine("tests", results.totalTests(), results.totalFailures(), results.failed());
-    builder.append(totalStatsLine);
-
-    var totalAssertsStatsLine =
-        makeStatsLine(
-            "asserts", results.totalAsserts(), results.totalAssertsFailed(), results.failed());
-    builder.append(", ").append(totalAssertsStatsLine);
-
-    builder.append("\n");
+    writer.append(builder.toString());
+  }
 
-    writer.append(builder);
+  public void summarize(List<TestResults> allTestResults, Writer writer) throws IOException {
+    var totalTests = 0;
+    var totalFailedTests = 0;
+    var totalAsserts = 0;
+    var totalFailedAsserts = 0;
+    var isFailed = false;
+    var isExampleWrittenFailure = true;
+    for (var testResults : allTestResults) {
+      if (!isFailed) {
+        isFailed = testResults.failed();
+      }
+      if (testResults.failed()) {
+        isExampleWrittenFailure = testResults.isExampleWrittenFailure() & isExampleWrittenFailure;
+      }
+      totalTests += testResults.totalTests();
+      totalFailedTests += testResults.totalFailures();
+      totalAsserts += testResults.totalAsserts();
+      totalFailedAsserts += testResults.totalAssertsFailed();
+    }
+    var builder = new TextFormattingStringBuilder(useColor);
+    if (isFailed && isExampleWrittenFailure) {
+      builder.append(totalFailedTests).append(" examples written");
+    } else {
+      makeStatsLine(builder, "tests", totalTests, totalFailedTests, isFailed);
+      builder.append(", ");
+      makeStatsLine(builder, "asserts", totalAsserts, totalFailedAsserts, isFailed);
+    }
+    builder.appendLine();
+    writer.append(builder.toString());
   }
 
-  private void reportResults(TestSectionResults section, StringBuilder builder) {
+  private void reportResults(TestSectionResults section, TextFormattingStringBuilder builder) {
     if (!section.results().isEmpty()) {
       builder.append("  ").append(section.name()).append("\n");
-
-      StringUtils.joinToStringBuilder(
-          builder, section.results(), "\n", res -> reportResult(res, builder));
+      builder.join(section.results(), "\n", res -> reportResult(res, builder));
       builder.append("\n");
```

Comment:
It's a utility method for string building; seems quite related?

For the sake of momentum, I'll just make this a private method in `SimpleReport`.

Reply:
I think there's a reason `java.lang.StringBuilder` doesn't have methods that are specific to collection types.

I agree that the `join` has _n - 1_ delimiters for any collection, instead of _n_, but you'll notice I also removed the `append("\n")` before. I'm saying writing this as a join wins nothing (I even think the `for` formulation is easier to read), but adds API surface.